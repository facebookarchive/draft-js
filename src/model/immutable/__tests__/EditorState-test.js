/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @oncall draft_js
 */

'use strict';

const CharacterMetadata = require('CharacterMetadata');
const ContentBlock = require('ContentBlock');
const ContentState = require('ContentState');
const DraftModifier = require('DraftModifier');
const EditorState = require('EditorState');
const RichTextEditorUtil = require('RichTextEditorUtil');
const {BOLD, ITALIC} = require('SampleDraftInlineStyle');
const SelectionState = require('SelectionState');

const Immutable = require('immutable');

const {List, Repeat} = Immutable;

class Decorator {}
Decorator.prototype.getDecorations = jest.fn();

const DEFAULT_SELECTION = {
  anchorKey: 'a',
  anchorOffset: 0,
  focusKey: 'a',
  focusOffset: 0,
  isBackward: false,
};

const collapsedSelection = new SelectionState(DEFAULT_SELECTION);
const rangedSelection = new SelectionState({
  ...DEFAULT_SELECTION,
  focusOffset: 1,
});

const plainBlock = new ContentBlock({
  key: 'a',
  text: 'Arsenal',
});
const boldBlock = new ContentBlock({
  key: 'b',
  text: 'Burnley',
  characterList: List(Repeat(CharacterMetadata.create({style: BOLD}), 7)),
});
const boldA = List(Repeat('x', boldBlock.getLength()));
const emptyBlockA = new ContentBlock({
  key: 'emptyA',
  text: '',
});
const emptyBlockB = new ContentBlock({
  key: 'emptyB',
  text: '',
});
const italicBlock = new ContentBlock({
  key: 'c',
  text: 'Chelsea',
  characterList: List(Repeat(CharacterMetadata.create({style: ITALIC}), 7)),
});

const getSampleEditorState = (type, decorator) => {
  switch (type) {
    case 'DECORATED':
      return EditorState.createWithContent(
        ContentState.createFromBlockArray([boldBlock, italicBlock]),
        decorator,
      );
    case 'MULTI_BLOCK':
      return EditorState.createWithContent(
        ContentState.createFromBlockArray([
          emptyBlockA,
          emptyBlockB,
          boldBlock,
        ]),
      );
    case 'UNDECORATED':
    default:
      return EditorState.createWithContent(
        ContentState.createFromBlockArray([
          plainBlock,
          boldBlock,
          emptyBlockA,
          emptyBlockB,
          italicBlock,
        ]),
      );
  }
};

const UNDECORATED_STATE = getSampleEditorState('UNDECORATED');

const MULTI_BLOCK_STATE = getSampleEditorState('MULTI_BLOCK');

const assertGetCurrentInlineStyle = (selection, state = UNDECORATED_STATE) => {
  const editorState = EditorState.acceptSelection(state, selection);
  expect(editorState.getCurrentInlineStyle().toJS()).toMatchSnapshot();
};

beforeEach(() => {
  Decorator.prototype.getDecorations.mockClear();
  Decorator.prototype.getDecorations.mockImplementation((v, c) => {
    return v === boldBlock ? boldA : List(Repeat(undefined, v.getLength()));
  });
});

test('uses right of the caret at document start', () => {
  assertGetCurrentInlineStyle(collapsedSelection);
});

test('uses left of the caret, at position `1+`', () => {
  assertGetCurrentInlineStyle(
    collapsedSelection.merge({
      anchorOffset: 1,
      focusOffset: 1,
    }),
  );
});

test('uses right of the caret at offset `0` within document', () => {
  assertGetCurrentInlineStyle(
    collapsedSelection.merge({
      anchorKey: 'b',
      focusKey: 'b',
    }),
  );
});

test('uses previous block at offset `0` within empty block', () => {
  assertGetCurrentInlineStyle(
    collapsedSelection.merge({
      anchorKey: 'emptyA',
      focusKey: 'emptyA',
    }),
  );
});

test('looks upward through empty blocks to find a character with collapsed selection', () => {
  assertGetCurrentInlineStyle(
    collapsedSelection.merge({
      anchorKey: 'emptyB',
      focusKey: 'emptyB',
    }),
  );
});

test('does not discard style override when changing block type', () => {
  let editor = EditorState.createEmpty();

  editor = RichTextEditorUtil.toggleInlineStyle(editor, 'BOLD');
  editor = RichTextEditorUtil.toggleBlockType(editor, 'test-block');

  expect(editor.getCurrentInlineStyle().toJS()).toMatchSnapshot();
});

test('does not discard style override when adjusting depth', () => {
  let editor = EditorState.createEmpty();

  editor = RichTextEditorUtil.toggleInlineStyle(editor, 'BOLD');
  editor = RichTextEditorUtil.onTab({preventDefault: () => {}}, editor);

  expect(editor.getCurrentInlineStyle().toJS()).toMatchSnapshot();
});

test('does not discard style override when splitting block', () => {
  let editor = EditorState.createEmpty();

  editor = RichTextEditorUtil.toggleInlineStyle(editor, 'BOLD');
  editor = EditorState.push(
    editor,
    DraftModifier.splitBlock(editor.getCurrentContent(), editor.getSelection()),
    'split-block',
  );

  expect(editor.getCurrentInlineStyle().toJS()).toMatchSnapshot();
});

test('uses right of the start for blocks with text', () => {
  assertGetCurrentInlineStyle(rangedSelection.set('focusKey', 'b'));
});

test('uses left of the start if starting at end of block', () => {
  const blockB = UNDECORATED_STATE.getCurrentContent().getBlockForKey('b');
  assertGetCurrentInlineStyle(
    collapsedSelection.merge({
      anchorKey: 'b',
      anchorOffset: blockB.getLength(),
      focusKey: 'c',
      focusOffset: 3,
    }),
  );
});

test('looks upward through empty blocks to find a character', () => {
  assertGetCurrentInlineStyle(
    rangedSelection.merge({
      anchorKey: 'emptyA',
      anchorOffset: 0,
      focusKey: 'c',
      focusOffset: 3,
    }),
  );
});

test('falls back to no style if in empty block at document start', () => {
  assertGetCurrentInlineStyle(
    new SelectionState({
      anchorKey: 'emptyA',
      anchorOffset: 0,
      focusKey: 'c',
      focusOffset: 3,
      isBackward: false,
    }),
    MULTI_BLOCK_STATE,
  );
});

test('must set a new decorator', () => {
  const decorator = new Decorator();
  const editorState = getSampleEditorState('DECORATED', decorator);
  const boldB = List(Repeat('y', boldBlock.getLength()));

  expect(decorator.getDecorations.mock.calls.length).toMatchSnapshot();

  Decorator.prototype.getDecorations.mockImplementation((v, c) => {
    return v === boldBlock ? boldB : List(Repeat(undefined, v.getLength()));
  });

  class NextDecorator {}
  NextDecorator.prototype.getDecorations = jest.fn();

  const newDecorator = new NextDecorator();

  NextDecorator.prototype.getDecorations.mockImplementation((v, c) => {
    return v === boldBlock ? boldB : List(Repeat(undefined, v.getLength()));
  });

  const withNewDecorator = EditorState.set(editorState, {
    decorator: newDecorator,
  });

  expect(withNewDecorator !== editorState).toMatchSnapshot();

  // Twice for the initial tree map generation, then twice more for
  // filter comparison.
  expect(decorator.getDecorations.mock.calls.length).toMatchSnapshot();

  // Twice for filter comparison, once for tree generation since one
  // block has the same decoration list and is filtered out.
  expect(newDecorator.getDecorations.mock.calls.length).toMatchSnapshot();

  expect(withNewDecorator.getDecorator() === newDecorator).toMatchSnapshot();

  // Preserve block trees that had the same decorator list.
  expect(
    editorState.getBlockTree(boldBlock.getKey()) ===
      withNewDecorator.getBlockTree(boldBlock.getKey()),
  ).toMatchSnapshot();

  expect(
    editorState.getBlockTree(italicBlock.getKey()) !==
      withNewDecorator.getBlockTree(italicBlock.getKey()),
  ).toMatchSnapshot();
});

test('must call decorator with correct argument types and order', () => {
  const decorator = new Decorator();
  getSampleEditorState('DECORATED', decorator);
  expect(decorator.getDecorations.mock.calls.length).toMatchSnapshot();
});

test('must correctly remove a decorator', () => {
  const decorator = new Decorator();
  const editorState = getSampleEditorState('DECORATED', decorator);
  const withNewDecorator = EditorState.set(editorState, {decorator: null});

  expect(withNewDecorator !== editorState).toMatchSnapshot();
  expect(decorator.getDecorations.mock.calls.length).toMatchSnapshot();
  expect(withNewDecorator.getDecorator()).toMatchSnapshot();
});
