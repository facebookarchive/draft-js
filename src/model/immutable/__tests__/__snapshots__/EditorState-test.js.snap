// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not discard style override when adjusting depth 1`] = `
Array [
  "BOLD",
]
`;

exports[`does not discard style override when changing block type 1`] = `
Array [
  "BOLD",
]
`;

exports[`does not discard style override when splitting block 1`] = `
Array [
  "BOLD",
]
`;

exports[`falls back to no style if in empty block at document start 1`] = `Array []`;

exports[`looks through rest backward-selected blocks to find a character when anchor block is empty 1`] = `
Array [
  "BOLD",
]
`;

exports[`looks through rest selected blocks to find a character when anchor block is empty 1`] = `
Array [
  "ITALIC",
]
`;

exports[`looks upward through empty blocks to find a character 1`] = `
Array [
  "BOLD",
]
`;

exports[`looks upward through empty blocks to find a character with collapsed selection 1`] = `
Array [
  "BOLD",
]
`;

exports[`must call decorator with correct argument types and order 1`] = `2`;

exports[`must correctly remove a decorator 1`] = `true`;

exports[`must correctly remove a decorator 2`] = `2`;

exports[`must correctly remove a decorator 3`] = `null`;

exports[`must set a new decorator 1`] = `2`;

exports[`must set a new decorator 2`] = `true`;

exports[`must set a new decorator 3`] = `4`;

exports[`must set a new decorator 4`] = `3`;

exports[`must set a new decorator 5`] = `true`;

exports[`must set a new decorator 6`] = `true`;

exports[`must set a new decorator 7`] = `true`;

exports[`uses left of the caret, at position \`1+\` 1`] = `Array []`;

exports[`uses left of the start if starting at end of block 1`] = `
Array [
  "BOLD",
]
`;

exports[`uses previous block at offset \`0\` within empty block 1`] = `
Array [
  "BOLD",
]
`;

exports[`uses right of the caret at document start 1`] = `Array []`;

exports[`uses right of the caret at offset \`0\` within document 1`] = `
Array [
  "BOLD",
]
`;

exports[`uses right of the start for blocks with text 1`] = `Array []`;
